proxy/grpcproxy/lease.go

```diff
@@ -46,42 +94,209 @@ func (lp *leaseProxy) LeaseTimeToLive(ctx context.Context, rr *pb.LeaseTimeToLiv
 	return pb.NewLeaseClient(conn).LeaseTimeToLive(ctx, rr)
 }
 
-func (lp *leaseProxy) LeaseKeepAlive(stream pb.Lease_LeaseKeepAliveServer) error {
-	conn := lp.client.ActiveConnection()
-	ctx, cancel := context.WithCancel(stream.Context())
-	lc, err := pb.NewLeaseClient(conn).LeaseKeepAlive(ctx)
-	if err != nil {
-		cancel()
-		return err
+func (lp *leaseProxy) LeaseKeepAlive(stream pb.Lease_LeaseKeepAliveServer) (err error) {
+	fmt.Printf("LeaseKeepAlive() new stream %v \n", stream)
+	lp.mu.Lock()
+	select {
+	case <-lp.ctx.Done():
+		lp.mu.Unlock()
+		return
+	default:
+		lp.wg.Add(1)
+	}
+	lp.mu.Unlock()
+
+	cctx, ccancel := context.WithCancel(stream.Context())
+	lps := leaseProxyStream{
+		stream:          stream,
+		client:          lp.client,
+		keepAliveLeases: make(map[int64]chan struct{}),
+		lrpCh:           make(chan *pb.LeaseKeepAliveResponse, 16),
+		ctx:             cctx,
+	}
+
+	var leaderc <-chan struct{}
+	if md, ok := metadata.FromContext(stream.Context()); ok {
+		v := md[rpctypes.MetadataRequireLeaderKey]
+		if len(v) > 0 && v[0] == rpctypes.MetadataHasLeader {
+			leaderc = lp.lostLeaderNotify()
+		}
 	}
 
+	stopc := make(chan struct{}, 3)
+	errCh := make(chan error, 1)
+
 	go func() {
-		// Cancel the context attached to lc to unblock lc.Recv when
-		// this routine returns on error.
-		defer cancel()
-
-		for {
-			// stream.Recv will be unblock when the loop in the parent routine
-			// returns on error.
-			rr, err := stream.Recv()
-			if err != nil {
-				return
-			}
-			err = lc.Send(rr)
-			if err != nil {
-				return
-			}
+		defer func() { stopc <- struct{}{} }()
+		errCh <- lps.recvLoop()
+	}()
+
+	// watches all the listeningFromServer() go routines
+	go func() {
+		defer func() { stopc <- struct{}{} }()
+		lps.sendLoop()
+	}()
+
+	// tears down LeaseKeepAlive stream if leader goes down or entire leaseProxy is terminated.
+	go func() {
+		defer func() { stopc <- struct{}{} }()
+		select {
+		case <-leaderc:
+		case <-cctx.Done():
+		case <-lp.ctx.Done():
 		}
 	}()
 
-	for {
-		rr, err := lc.Recv()
-		if err != nil {
-			return err
+	<-stopc
+	ccancel()
+
+	go func() {
+		<-stopc
+		<-stopc
+		lps.close()
+		lp.wg.Done()
+	}()
+
+	select {
+	case <-leaderc:
+		return rpctypes.ErrNoLeader
+	case err = <-errCh:
+		return err
+	default:
+		return lp.ctx.Err()
+	}
+}
+
+type leaseProxyStream struct {
+	stream pb.Lease_LeaseKeepAliveServer
+	client *clientv3.Client
+
+	wg sync.WaitGroup
+	// mu protects keepAliveLeases
+	mu sync.RWMutex
+	// keepAliveLeases map keeps track of on-going lease renewal IDs and corresponding chans.
+	// the chan tracks number of keepalive requests received for a given lease ID.
```

Comment:
channels shouldn't track the number of requests received-- they're for messaging; counting doesn't work because no matter how much it's buffered, there's a chance it'll fill up and block